### PR TITLE
New algorithm: Spanning Edge Centrality Approximation

### PR DESCRIPTION
**Pull request by angriman**
This PR brings a parallel implementation of the [Hayashi et al. algorithm](https://www.ijcai.org/Proceedings/16/Papers/525.pdf) for spanning edge centrality approximation.

To make the algorithm more efficient, this PR also includes a new function `Graph::getIthNeighborWithId` that returns the ith neighbor of a node `u` together with the id of the edge `(u, v)`. This is faster than using `Graph::getIthNeighbor` and retrieving the edge id with `Graph::edgeId`.

### CODE REVIEW

File: networkit/test/test_algorithms.py [hunk 1]
```diff
@@ -15,22 +15,22 @@ def checkCovers(self, c1, c2):
...
-		#self.L = readGraph("PGPgiantcompo.graph", Format.METIS) #without self-loops
-		#self.LL = readGraph("PGPConnectedCompoLoops.gml", Format.GML) #with self-loops sprinkled in
-		self.L = readGraph("input/looptest1.gml", Format.GML) #without self-loops
-		self.LL = readGraph("input/looptest2.gml", Format.GML) #with self-loops sprinkled in
+		#self.L = nk.readGraph("PGPgiantcompo.graph", nk.Format.METIS) #without self-loops
+		#self.LL = nk.readGraph("PGPConnectedCompoLoops.gml", nk.Format.GML) #with self-loops sprinkled in
+		self.L = nk.readGraph("input/looptest1.gml", nk.Format.GML) #without self-loops
+		self.LL = nk.readGraph("input/looptest2.gml", nk.Format.GML) #with self-loops sprinkled in
 
 	def test_centrality_Betweenness(self):
```
[the remark this answers is not in the document]
Comment (angriman) [reply networkit/test/test_algorithms.py]: That's true, I added another commit that refactors that.

**Comment by avdgrinten** [comment]
Regarding Python tests: I don't see a need to exhaustively test Python interfaces. The notebooks are also execute on CI. If you just want to make sure that the Python version runs, adding it to the notebook is sufficient.

**Comment by angriman** [comment]
Thank you for the review @manpen, I've addressed the comments.

Regarding Python tests, I agree that they are not needed if the Python interface is already tested in the notebooks. We could write Python tests only for the classes that have some additional functionalities that are not shown in the notebooks.

**Comment by fabratu** [comment]
@angriman Can you resolve the conflicts? Apart from that I think all change requests have been taken care of. @manpen Can you approve/deny?

**Comment by fabratu** [comment]
Thx and agreed - the faster implementation should be properly integrated into NetworKit in order to leverage the improvements where applicable. Is this faster random-generator already present somewhere else in the code? Also: There are still several occasions in the code, where `Aux::Random` is still called. Is this intended?

File: include/networkit/centrality/ApproxSpanningEdge.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+#ifndef NETWORKIT_CENTRALITY_APPROX_SPANNING_EDGE_HPP_
+#define NETWORKIT_CENTRALITY_APPROX_SPANNING_EDGE_HPP_
+
+#include <vector>
+
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+#define PCG32_MULT 0x5851f42d4c957f2dULL
```
Comment (manpen) [inline include/networkit/centrality/ApproxSpanningEdge.hpp:19]: Please use a appropriately scoped constexpr

**Comment by manpen** [comment]
Thanks a lot to @fabratu for taking over. 

Just out of curiosity: how much of an impact does integrating PCG make? The reason I'm asking is that some time ago, I tried to reproducing the acclaimed speed-ups, and never saw substantial increases compared to libstd++ MT in practice. Is it due to the reduced state space here?

File: include/networkit/centrality/ApproxSpanningEdge.hpp (hunk 2 above)
Comment (angriman) [reply include/networkit/centrality/ApproxSpanningEdge.hpp]: Good point, updated.

**Comment by angriman** [comment]
The comparison that we did was against `Graph::randomNeighbor`, which generates a uniform_int_distribution each time it is called with some upper/lower bound. I guess that was the main reason that slowed it down. However, to use the standard MT one should also generate a uniform_int_distribution for each vertex, which can be avoided with PCG.

File: networkit/cpp/centrality/ApproxSpanningEdge.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    visitedNodes.resize(
+        omp_get_max_threads(),
+        std::vector<NodeStatus>(G.upperNodeIdBound(), NodeStatus::NOT_IN_COMPONENT));
+    edgeScores.resize(omp_get_max_threads(), std::vector<count>(G.upperEdgeIdBound(), 0));
+    parents.resize(omp_get_max_threads(), std::vector<node>(G.upperNodeIdBound(), none));
+    parentsEdgeIds.resize(omp_get_max_threads(), std::vector<edgeid>(G.upperNodeIdBound(), none));
+    generators.reserve(omp_get_max_threads());
+    for (omp_index i = 0; i < static_cast<omp_index>(omp_get_max_threads()); ++i) {
+        generators.emplace_back((Aux::Random::integer(), Aux::Random::integer()));
+        generators.back().seed(Aux::Random::integer());
```
[the remark this answers is not in the document]
Comment (angriman) [reply networkit/cpp/centrality/ApproxSpanningEdge.cpp:37]: Good point, let me fix that.

**Comment by manpen** [comment]
I would not expect that creating `uniform_int_distritbution` to cause a significant slow-down; after all in `stdlibc++` the constructor stores only two words without computation. Using `randomNeighbor` however involves memory accesses, and retrieves from `Aux::Random::getURNG()` involving checks for a static instance. At one point, we measured that a tight loop `for(...) {Aux::Random::getInteger();}` is roughly 30% slower than `auto& gen = Aux::Random::getURNG(); for(...) {std::uniform_int_distr{a,b}(gen)`. Are you convinced that PCG has to be added/can measure by using `auto&genenerator=Aux::Random::getURNG()` in `sampleUST`;?

Earlier in this PR there was a discussion about adding a Notebook. Is this still planned?

Both comments are intended as open questions, and wont block merging if you do not have capacity to proceed.

**Comment by angriman** [comment]
The only reason I'd add PCG is to avoid that 30% slowdown you mentioned, I did not know exactly what caused it. If we can avoid it with `auto& gen = Aux::Random::getURNG(); for(...) {std::uniform_int_distr{a,b}(gen)` I'd prefer this solution than using PCG.

**Comment by angriman** [comment]
Thank you all for the reviews. @manpen I added an example in the Centrality notebook, and apparently for this PR your approval is needed.

**Pull request merged**